packages/block-library/src/navigation/edit/navigation-menu-selector.js

```diff
@@ -5,82 +5,172 @@ import {
+
 	actionLabel = actionLabel || createActionLabel;
 
 	const { menus: classicMenus } = useNavigationEntities();
 
 	const {
 		navigationMenus,
+		hasResolvedNavigationMenus,
+		isNavigationMenuResolved,
 		canUserCreateNavigationMenu,
-		canUserUpdateNavigationMenu,
 		canSwitchNavigationMenu,
 	} = useNavigationMenu();
 
+	const [ currentTitle ] = useEntityProp(
+		'postType',
+		'wp_navigation',
+		'title'
+	);
+
 	const menuChoices = useMemo( () => {
 		return (
 			navigationMenus?.map( ( { id, title } ) => {
 				const label = decodeEntities( title.rendered );
+				if ( id === currentMenuId && ! isCreatingMenu ) {
+					setSelectorLabel( currentTitle );
+					setEnableOptions(
+						( canSwitchNavigationMenu ||
+							canUserCreateNavigationMenu ) &&
+							hasResolvedNavigationMenus &&
+							( ! isCreatingMenu ||
+								createNavigationMenuIsSuccess )
+					);
+				}
 				return {
 					value: id,
 					label,
 					ariaLabel: sprintf( actionLabel, label ),
 				};
 			} ) || []
 		);
-	}, [ navigationMenus ] );
+	}, [
+		currentTitle,
+		currentMenuId,
+		navigationMenus,
+		createNavigationMenuIsSuccess,
+		isNavigationMenuResolved,
+		hasResolvedNavigationMenus,
+	] );
 
 	const hasNavigationMenus = !! navigationMenus?.length;
 	const hasClassicMenus = !! classicMenus?.length;
 	const showNavigationMenus = !! canSwitchNavigationMenu;
 	const showClassicMenus = !! canUserCreateNavigationMenu;
-	const hasManagePermissions =
-		canUserCreateNavigationMenu || canUserUpdateNavigationMenu;
-
-	// Show the selector if:
-	// - has switch or create permissions and there are block or classic menus.
-	// - user has create or update permissions and component should show the menu actions.
-	const showSelectMenus =
-		( ( canSwitchNavigationMenu || canUserCreateNavigationMenu ) &&
-			( hasNavigationMenus || hasClassicMenus ) ) ||
-		( hasManagePermissions && showManageActions );
-
-	if ( ! showSelectMenus ) {
-		return null;
+
+	useEffect( () => {
+		if ( ! hasResolvedNavigationMenus ) {
+			setSelectorLabel( __( 'Loading …' ) );
+		} else if ( hasNavigationMenus && ! currentMenuId ) {
+			setSelectorLabel( __( 'Select a menu' ) );
+			setEnableOptions(
+				( canSwitchNavigationMenu || canUserCreateNavigationMenu ) &&
+					hasResolvedNavigationMenus &&
+					( ! isCreatingMenu || createNavigationMenuIsSuccess )
+			);
+		} else if ( ! hasNavigationMenus && hasResolvedNavigationMenus ) {
+			setSelectorLabel( __( 'No menus. Create one?' ) );
+			setEnableOptions(
+				( canSwitchNavigationMenu || canUserCreateNavigationMenu ) &&
+					hasResolvedNavigationMenus &&
+					( ! isCreatingMenu || createNavigationMenuIsSuccess )
+			);
+		} else if ( hasResolvedNavigationMenus && currentMenuId === null ) {
+			setSelectorLabel( __( 'Select another menu' ) );
+			setEnableOptions(
+				( canSwitchNavigationMenu || canUserCreateNavigationMenu ) &&
+					hasResolvedNavigationMenus &&
+					( ! isCreatingMenu || createNavigationMenuIsSuccess )
+			);
```

Comment:
I wonder if some of these could be refactored to well named variables? At the moment one has to parse and understand all these conditionals. Is there a way to make it easy to understand at a glance?

Reply:
Done.